waiter/src/waiter/scheduler/marathon.clj

```diff
@@ -395,11 +419,150 @@
+  [marathon-scheduler current-time trigger-timeout]
+  (let [service-id->out-of-sync-state (retrieve-service-id->out-of-sync-state marathon-scheduler)
+        service-id->instance-count-map (retrieve-out-of-sync-apps marathon-scheduler)
+        out-of-sync-service-ids (-> service-id->instance-count-map keys set)
+        trigger-time (t/minus current-time trigger-timeout)
+        trigger-service-ids (->> service-id->out-of-sync-state
+                                 (filter (fn out-of-sync-trigger-predicate
+                                           [[service-id {:keys [data last-modified-time]}]]
+                                           (and (= data (service-id->instance-count-map service-id))
+                                                (t/before? last-modified-time trigger-time))))
+                                 (map first)
+                                 set)]
+    (log/info (count out-of-sync-service-ids) "service(s) have out-of-sync deployments:" out-of-sync-service-ids)
+    (log/info (count trigger-service-ids) "service(s) qualify for sync deployments:" trigger-service-ids)
+    {:service-id->instance-count-map service-id->instance-count-map
+     :service-id->out-of-sync-state (->> (set/difference out-of-sync-service-ids trigger-service-ids)
+                                         (pc/map-from-keys
+                                           (fn compute-out-of-sync-state [service-id]
+                                             (let [prev-state (service-id->out-of-sync-state service-id)
+                                                   prev-data (:data prev-state)
+                                                   curr-data (service-id->instance-count-map service-id)]
+                                               (if (= prev-data curr-data)
+                                                 prev-state
+                                                 {:data curr-data :last-modified-time current-time})))))
+     :trigger-service-ids trigger-service-ids}))
+
+(defn- process-out-of-sync-services!
+  "Determines the out-of-sync services and triggers sync deployments on those services if needed.
+   As a side-effect, it updates the service-id->out-of-sync-state in the scheduler."
+  [marathon-scheduler current-time  trigger-timeout]
+  (try
+    (let [{:keys [service-id->instance-count-map service-id->out-of-sync-state trigger-service-ids]}
+          (compute-out-of-sync-state marathon-scheduler current-time trigger-timeout)]
+      (doseq [service-id trigger-service-ids]
+        (->> (service-id->instance-count-map service-id)
+             (trigger-sync-deployment! marathon-scheduler service-id)))
+      (log/info "updating local service-id->out-of-sync-state")
+      (reset-service-id->out-of-sync-state! marathon-scheduler service-id->out-of-sync-state))
+    (catch Exception e
+      (log/error e "unable to process out-of-sync services"))))
```

Comment:
Does it makes sense to leave the value of `service-id->out-of-sync-state` untouched in the case of an error? Or would it be better to reset it to some other value?

Reply:
Now that `service-id->out-of-sync-state-store` is an agent I do this, but just modify the time in the state.